### PR TITLE
Support tensorrt int8 quantization

### LINKED ISSUE #533: Tensorrt int8 quantization function cannot be used

### PR DESCRIPTION
**Pull request by XueweiHan21**
Closes #533.

Signed-off-by: HanXuewei <han.xuewei@zte.com.cn>

### CODE REVIEW

**Comment by amadeus-zte** [comment]
Is there a readme on how to use calibratior?

File: model_compiler/src/model_compiler/calibrator.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+        self.data_size = trt.volume([self.batch_size, self.channel, self.height, self.width]) * trt.float32.itemsize
+        self.device_input = cuda.mem_alloc(self.data_size)
+
+    def _next_batch(self):
+        if self.current_batch < self.max_batch:
+            batch_files = self.images[self.current_batch * self.batch_size: (self.current_batch + 1) * self.batch_size]
+            batch_images = np.zeros((self.batch_size, self.channel, self.height, self.width), dtype=np.float32)
+            normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
+            img_transform = transforms.Compose([transforms.CenterCrop(self.width), transforms.ToTensor(), normalize, ])
```
Comment (NaNAGISaSA) [inline model_compiler/src/model_compiler/calibrator.py:49]: `normalize, `?

File: model_compiler/src/model_compiler/calibrator.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            for i, f in enumerate(batch_files):
+                img = Image.open(f)
+                if len(img.split()) != 3:
```
Comment (NaNAGISaSA) [inline model_compiler/src/model_compiler/calibrator.py:52]: If there is a broken image, you will use np.zeros to calibrate, is it ok?

File: model_compiler/src/model_compiler/calibrator.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            return np.ascontiguousarray(batch_images)
+        else:
+            return np.array([])
+
+    def get_batch_size(self):
+        return self.batch_size
+
+    def get_batch(self, names, p_str=None):
+        batch_imgs = self._next_batch()
+        if batch_imgs.size == 0 or batch_imgs.size != self.batch_size*self.channel*self.height*self.width:
```
Comment (NaNAGISaSA) [inline model_compiler/src/model_compiler/calibrator.py:69]: need blanks

File: model_compiler/src/model_compiler/compilers/onnx_model_to_tensorrt_model.py [hunk 4]
```diff
@@ -72,15 +78,23 @@ def compile_source(source: OnnxModel, config: Config) -> TensorRTModel:
 
             builder_config.add_optimization_profile(optimization_profile)
 
-        if config.int8_calibrator:
-            builder_config.int8_calibrator = config.int8_calibrator
-
-        if config.enable_fp16:
+        if config.data_type == 'FP16':
             builder_config.set_flag(tensorrt.BuilderFlag.FP16)
+            builder_config.set_flag(tensorrt.BuilderFlag.STRICT_TYPES)
```
Comment (NaNAGISaSA) [inline model_compiler/src/model_compiler/compilers/onnx_model_to_tensorrt_model.py:83]: opposite with `config.enable_strict_types` when it is False

File: model_compiler/src/model_compiler/calibrator.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        self.batch_size = batch_size
+        self.channel = sorted(input_shape[1:])[0]
+        self.width = sorted(input_shape[1:])[1]
+        self.height = sorted(input_shape[1:])[2]
+
+        self.images = _load_calibration_data(calibration_dataset)
+        self.current_batch = 0
+        self.max_batch = len(self.images)//self.batch_size
+
+        self.data_size = trt.volume([self.batch_size, self.channel, self.height, self.width]) * trt.float32.itemsize
```
Comment (NaNAGISaSA) [inline model_compiler/src/model_compiler/calibrator.py:41]: `self.data_size` member is no need

**Review by NaNAGISaSA — CHANGES_REQUESTED** [review]

File: docker-images/dockerfiles/model-compiler.dockerfile [hunk 6]
```diff
@@ -58,16 +58,19 @@ RUN apt-get update && \
     find /var/lib/apt/lists -delete &&\
     ln -s /usr/local/cuda-"$CUDA_VERSION" /usr/local/cuda
 
-COPY --from=builder /src/dist/*.whl /tmp/model-compiler-package/
+RUN apt-get update && apt-get install -y git gcc python3-dev python3-pip libxml2-dev libxslt1-dev zlib1g-dev g++
```
[the remark this answers is not in the document]
Comment (XueweiHan21) [reply docker-images/dockerfiles/model-compiler.dockerfile:61]: These packages are required when installing pycuda.

File: model_compiler/src/model_compiler/calibrator.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
+# Copyright 2021 ZTE corporation. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+import tensorrt as trt
+import pycuda.driver as cuda
+import pycuda.autoinit  # noqa # pylint: disable=unused-import
```
[the remark this answers is not in the document]
Comment (XueweiHan21) [reply model_compiler/src/model_compiler/calibrator.py:6]: It is needed, it needs to be initialized before using cuda quantitatively, refer to https://github.com/NVIDIA/TensorRT/blob/main/samples/python/int8_caffe_mnist/calibrator.py

File: model_compiler/src/model_compiler/calibrator.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+class MyCalibrator(trt.IInt8EntropyCalibrator2):
+
+    def __init__(self, calibration_dataset, batch_size, input_shapes):
+        trt.IInt8EntropyCalibrator2.__init__(self)
+
+        self.cache_file = None
+
+        input_shape = input_shapes[0]
+        self.batch_size = batch_size
+        self.channel = sorted(input_shape[1:])[0]
```
[the remark this answers is not in the document]
Comment (XueweiHan21) [reply model_compiler/src/model_compiler/calibrator.py:33]: The default calibrator is only a reference for general conditions.

File: model_compiler/src/model_compiler/calibrator.py (hunk 1 above)
Comment (XueweiHan21) [reply model_compiler/src/model_compiler/calibrator.py:49]: The usage of transforms.Compose(), refer to the source https://pytorch.org/vision/stable/transforms.html

**Comment by XueweiHan21** [comment]
Submit in a new pull request: https://github.com/Adlik/Adlik/pull/548

**Pull request closed without merging**